build-tools/src/main/java/org/elasticsearch/gradle/plugin/GenerateTestBuildInfoTask.java

```diff
@@ -0,0 +1,294 @@
+                    sb.append(entry.getValue());
+                    sb.append("\"\n        },\n");
+                }
+                writer.write(sb.substring(0, sb.length() - 2));
+            }
+            writer.write("\n    ]\n}\n");
+        }
+    }
+
+    // build the association for a class to a module;
+    // there are different methods for finding these depending on if the
+    // classpath entry is a jar or a directory
+    private Map<String, String> buildClassesToModules() throws IOException {
+        Map<String, String> classesToModules = new HashMap<>();
+        for (File file : getCodeLocations().get().getFiles()) {
+            if (file.exists()) {
+                if (file.getName().endsWith(".jar")) {
+                    extractFromJar(file, classesToModules);
+                } else if (file.isDirectory()) {
+                    extractFromDirectory(file, classesToModules);
+                } else {
+                    throw new IllegalArgumentException("unrecognized classpath entry: " + file);
+                }
+            }
+        }
+        return classesToModules;
+    }
+
+    // find the first class and module when the class path entry is a jar
+    private void extractFromJar(File file, Map<String, String> classesToModules) throws IOException {
+        try (JarFile jarFile = new JarFile(file)) {
+            String className = extractClassNameFromJar(jarFile);
+            String moduleName = extractModuleNameFromJar(file, jarFile);
+
+            if (className != null) {
+                classesToModules.put(className, moduleName);
+            }
+        }
+    }
+
+    // look through the jar to find the first unique class that isn't
+    // in META-INF (those may not be unique) and isn't module-info.class
+    // (which is also not unique) and avoid anonymous classes
+    private String extractClassNameFromJar(JarFile jarFile) {
+        return jarFile.stream()
+            .filter(
+                je -> je.getName().startsWith("META-INF") == false
+                    && je.getName().equals("module-info.class") == false
+                    && je.getName().contains("$") == false
+                    && je.getName().endsWith(".class")
+            )
+            .findFirst()
+            .map(ZipEntry::getName)
+            .orElse(null);
+    }
+
+    // look through the jar for the module name with
+    // each step commented inline
+    private String extractModuleNameFromJar(File file, JarFile jarFile) throws IOException {
+        String moduleName = null;
+
+        // if the jar is multi-release, there will be a set versions
+        // under the path META-INF/versions/<version number>;
+        // each version will have its own module-info.class if this is a modular jar;
+        // look for the module name in the module-info from the latest version
+        // fewer than or equal to the current JVM version
+        if (jarFile.isMultiRelease()) {
```

Comment:
Do we need this complexity? I would expect the module name to be the same across all module-info files even in a MR jar. Assuming this is the case, can we just grab the first one regardless?

Reply:
Probably? I was ensuring we followed the spec, though. I would guess we could run into issues with older 3rd party libs potentially.

Reply:
> Probably? I was ensuring we followed the spec, though. I would guess we could run into issues with older 3rd party libs potentially.

By this you mean jars targgetting pre-Java 9 so they have the module-info in an MR folder?